### PR TITLE
update: MSRV from 1.56.1 to 1.58.1

### PR DESCRIPTION
**Pull request by Velfi**
## Motivation and Context

#1318 

## Description

This PR bumps the SDK's MSRV from 1.56.1 to 1.58.1 per our MSRV "2 versions behind" policy. It also fixes many issues highlighted by the latest clippy lints.

## Testing

ran existing tests

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: CHANGELOG.next.toml [hunk 1]
```diff
@@ -10,3 +10,15 @@
 # references = ["smithy-rs#920"]
 # meta = { "breaking" = false, "tada" = false, "bug" = false }
 # author = "rcoh"
+
+[[aws-sdk-rust]]
+message = "Bump [MSRV](https://github.com/awslabs/aws-sdk-rust#supported-rust-versions-msrv) from 1.56.1 to 1.58.1 per our \"two versions behind\" policy."
+references = ["smithy-rs#1318"]
+meta = { "breaking" = false, "tada" = false, "bug" = false }
```
Comment (jdisanti) [inline CHANGELOG.next.toml:17]: Seems like this change should be marked as breaking

File: aws/rust-runtime/aws-config/src/ecs.rs [hunk 2]
```diff
@@ -131,6 +131,7 @@ impl ProvideCredentials for EcsCredentialsProvider {
 
 /// Inner Provider that can record failed configuration state
 #[derive(Debug)]
+#[allow(clippy::large_enum_variant)]
```
Comment (jdisanti) [inline aws/rust-runtime/aws-config/src/ecs.rs:134]: How big is it?

**Review by jdisanti — APPROVED** [review]

File: aws/rust-runtime/aws-config/src/ecs.rs (hunk 2 above)
Comment (Velfi) [reply aws/rust-runtime/aws-config/src/ecs.rs:134]: 432 bytes, the next largest is 56 bytes. Clippy recommends boxing it.

File: CHANGELOG.next.toml (hunk 1 above)
Comment (Velfi) [inline CHANGELOG.next.toml:17]:
```suggestion
meta = { "breaking" = true, "tada" = false, "bug" = false }
```

File: CHANGELOG.next.toml [hunk 3]
```diff
@@ -10,3 +10,15 @@
...
+[[aws-sdk-rust]]
+message = "Bump [MSRV](https://github.com/awslabs/aws-sdk-rust#supported-rust-versions-msrv) from 1.56.1 to 1.58.1 per our \"two versions behind\" policy."
+references = ["smithy-rs#1318"]
+meta = { "breaking" = true, "tada" = false, "bug" = false }
+author = "Velfi"
+
+[[smithy-rs]]
+message = "Bump [MSRV](https://github.com/awslabs/aws-sdk-rust#supported-rust-versions-msrv) from 1.56.1 to 1.58.1 per our \"two versions behind\" policy."
+references = ["smithy-rs#1318"]
+meta = { "breaking" = false, "tada" = false, "bug" = false }
```
Comment (Velfi) [inline CHANGELOG.next.toml:23]:
```suggestion
meta = { "breaking" = true, "tada" = false, "bug" = false }
```

**Pull request merged**